nservicebus/upgrades/5to6/distributor.md

```diff
@@ -0,0 +1,128 @@
+---
+title: Migrating the Distributor to use Sender Side Distribution
+tags:
+ - upgrade
+ - migration
+related:
+ - samples/scaleout/distributor-upgrade
+---
+
+The [distributor](/nservicebus/scalability-and-ha/distributor) has been deprecated and replaced by [sender-side distribution](/nservicebus/msmq/scalability-and-ha/sender-side-distribution.md) with NServiceBus Version 6.
+
+## Client-side distribution
+
+With sender-side distribution, senders are aware of scaled out receivers, allowing senders to distribute messages directly to the receivers instead of using an intermediary distributor.
+
+
+## General upgrade scenario
+
+This process aims to allow upgrade in production environments without missing any messages and minimal downtime. If a scaled-out endpoint only receives commands and does not subscribe to messages consider following the [Simple upgrade scenario](#simple-upgrade-scenario) instead.
+
+* Upgrade all endpoints that interact with the distributor to Version 6 first. At this stage, do not upgrade the workers to Version 6.
+  * upgrade all endpoints that send command messages to the Distributor endpoint to be distributed,
+  * upgrade all endpoints that send subscription messages to the Distributor, 
+  * upgrade all endpoints subscribing to events published by worker nodes.
+* Apply the following steps for each worker, one after another:
+  * Shut down the worker.
+  * [Upgrade to NServiceBus Version 6](#upgrade-endpoint-to-version-6).
+  * [Configure it to enlist it with the distributor](#enlist-version-6-endpoints-with-a-distributor).
+  * Start it again.
+* Configure [sender-side distribution](/nservicebus/msmq/scalability-and-ha/sender-side-distribution.md) for all endpoints sending commands or publishing events to the scaled out endpoint.
+* Detach the workers from the Distributor by applying the following steps to the instances enlisted to the Distributor. But *skip this step for at least one instance* to ensure some workers remain attached to the distributor.
+  * Shut down the worker.
```

Comment:
> As long as you have instances of the same endpoint BOTH running standalone AND as joined through their distributor AT THE SAME TIME, you will never drain the distributor queue.

that's not required in this step. As there is still at least one instance which will handle incoming messages. depending on the next step (migrate to worker or remove), there are different actions to take as described after this section.

> If you run workers for the same endpoint both joined and not joined to a distributor, you get duplicated subscriptions and thus never empty distributor queue. 

I don't think you do. As all the workers were upgraded, they will subscribe with (A@worker1, A) for detached and (A@Dist, A) for attached workers, the publisher will round-robbin events between those, not duplicate as both provide a logical endpoint name.